src/WebJobs.Script.Grpc/Channel/OrderedInvocationMessageDispatcher.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Grpc.Eventing;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Grpc;
+
+internal sealed class OrderedInvocationMessageDispatcher : IInvocationMessageDispatcher, IDisposable
+{
+    private readonly ILogger _logger;
+    private readonly string _invocationId;
+    private readonly Action<InboundGrpcEvent> _dispatchWithChannel;
+
+    // Separated these out for easier testing of the flow.
+    private readonly Action<InboundGrpcEvent> _dispatchWithThreadPool;
+
+    private Channel<InboundGrpcEvent> _channel;
+    private bool _isChannelInitialized = false;
+    private bool _invocationComplete = false;
+    private bool _disposed = false;
+
+    public OrderedInvocationMessageDispatcher(string invocationId, ILogger logger, Action<InboundGrpcEvent> processItem)
+        : this(invocationId, logger, processItem, processItem)
+    {
+    }
+
+    // For testing
+    internal OrderedInvocationMessageDispatcher(string invocationId, ILogger logger, Action<InboundGrpcEvent> processItemWithChannel,
+        Action<InboundGrpcEvent> processItemWithThreadPool)
+    {
+        _logger = logger;
+        _invocationId = invocationId;
+        _dispatchWithChannel = processItemWithChannel;
+        _dispatchWithThreadPool = processItemWithThreadPool;
+    }
+
+    // For testing
+    internal Channel<InboundGrpcEvent> MessageChannel => _channel;
+
+    private static Channel<InboundGrpcEvent> InitializeChannel() =>
+        Channel.CreateUnbounded<InboundGrpcEvent>(
+            new UnboundedChannelOptions
+            {
+                SingleReader = true,
+                SingleWriter = true
+            });
+
+    public void DispatchRpcLog(InboundGrpcEvent msg)
+    {
+        // The channel is only needed if we use any logs. Otherwise, ordering doesn't matter.
+        // This is not thread-safe, but it is always called in-order from the same thread, so no
+        // need for locking.
+        // We also do not want to initialize the channel if we're receiving an out-of-order RpcLog and
+        // the invocation is already done.
+        if (!_isChannelInitialized && !_invocationComplete)
```

Comment:
How do we know if a log is out of order? Are we just depending on if the invocation is complete or not?

Reply:
Yes. That's really the only "rule" -- once an invocation starts, we can get 2 messages; 
- one or more RpcLogs
- a single InvocationComplete

Once complete, we consider that invocation done and no more logs can be processed (and we dispose of Contexts, etc for that invocation). At that point, ordering doesn't matter. We're going to throw the log away anyway, so maintaining order isn't important.

And for the ordering of logs *during* and invocation, there is no real order. All we know is that we're getting them one-at-a-time from the worker. In the order that the worker determined to be correct. So if the worker is mangling the order (or incorrectly sending RpcLogs after an invocation is done), then there's not much we can do to "fix" it.

Reply:
I've updated the comment with this information...